tests/TestCase/Controller/ModulesControllerTest.php

```diff
@@ -0,0 +1,772 @@
+     * @var BEdita\SDK\BEditaClient
+     */
+    public $client;
+
+    /**
+     * Test request config
+     *
+     * @var array
+     */
+    public $defaultRequestConfig = [
+        'environment' => [
+            'REQUEST_METHOD' => 'GET',
+        ],
+        'get' => [],
+        'params' => [
+            'object_type' => 'documents',
+        ],
+    ];
+
+    /**
+     * Setup api client and auth
+     *
+     * @return void
+     */
+    private function setupApi() : void
+    {
+        $this->client = ApiClientProvider::getApiClient();
+        $adminUser = getenv('BEDITA_ADMIN_USR');
+        $adminPassword = getenv('BEDITA_ADMIN_PWD');
+        $response = $this->client->authenticate($adminUser, $adminPassword);
+        $this->client->setupTokens($response['meta']);
+    }
+
+    /**
+     * Setup controller to test with request config
+     *
+     * @param array|null $requestConfig
+     * @return void
+     */
+    protected function setupController(?array $requestConfig = []): void
+    {
+        $config = array_merge($this->defaultRequestConfig, $requestConfig);
+        $request = new ServerRequest($config);
+        $this->controller = new ModulesControllerSample($request);
+        $this->setupApi();
+        $this->createTestObject();
+    }
+
+    /**
+     * Test `initialize` method
+     *
+     * @covers ::initialize()
+     *
+     * @return void
+     */
+    public function testInitialize() : void
+    {
+        // Setup controller for test
+        $this->setupController(); // it already calls initialize, internally
+        static::assertEquals('documents', $this->controller->getObjectType());
+        static::assertEquals('documents', $this->controller->Modules->getConfig('currentModuleName'));
+        static::assertEquals('documents', $this->controller->Schema->getConfig('type'));
+    }
+
+    /**
+     * Test `index` method
+     *
+     * @covers ::index()
+     *
+     * @return void
+     */
+    public function testIndex() : void
+    {
+        // Setup controller for test
+        $this->setupController();
+
+        // do controller call
+        $result = $this->controller->index();
+
+        // verify response status code and type
+        static::assertNull($result);
+        static::assertEquals(200, $this->controller->response->statusCode());
+        static::assertEquals('text/html', $this->controller->response->type());
+
+        // verify expected vars in view
+        $this->assertExpectedViewVars(['objects', 'meta', 'links', 'types', 'properties']);
+    }
+
+    /**
+     * Test `descendants` method
+     *
+     * @covers ::descendants()
+     *
+     * @return void
+     */
+    public function testDescendants() : void
+    {
+        // Setup controller for test
+        $this->setupController();
+
+        // do controller call
+        $result = $this->controller->descendants();
+
+        // verify response status code and type
+        static::assertTrue(is_array($result));
```

Comment:
more meaningful test could be if `$result` is empty (on `documents`) or not empty (on `objects`)

Reply:
fair enough. coming soon ;)